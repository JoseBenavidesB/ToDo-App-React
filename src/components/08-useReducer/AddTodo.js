import React from 'react'
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({ handleAddToDo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: '',

    });

    const handleSubmit = (e) => {
        e.preventDefault();
        if ( description.trim().length <= 1) { return;}

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };
        
        handleAddToDo(newTodo);
        reset();
    }
    return (
        <>
            <h4> Add ToDo </h4> <hr/>

            <form onSubmit={ handleSubmit }>
                
                <input
                    type="text"
                    name="description"
                    className='form-control mb-0 m-auto'
                    placeholder="Write something"
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button 
                    type= "submit" 
                    className='btn btn-outline-primary mt-2 col-12 m-auto'> Add </button>

            </form>
        </>
    )
}
