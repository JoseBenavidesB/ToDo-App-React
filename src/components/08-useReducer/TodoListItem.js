import React from 'react'

export const TodoListItem = ({ todo, idx , handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        onClick={() => handleToggle(todo.id)}
        className={`${todo.done && "complete"}`}
      >
        {idx + 1}. {todo.desc}
      </p>

      <button
        // id= { todo.id }
        onClick={() => handleDelete(todo.id)}
        className="btn btn-danger m-2"
      >
        Delete
      </button>
    </li>
  );
}
