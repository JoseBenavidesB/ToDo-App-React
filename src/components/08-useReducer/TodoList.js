
import React from 'react'
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ toDos = [], handleToggle, handleDelete }) => {
  return (
    <ul className="list-group list-group-flush p-0">
        {toDos.map((todo, idx) => (
            <TodoListItem key={todo.id} todo={ todo } idx={ idx }  handleDelete={handleDelete} handleToggle={handleToggle}/>
        ))}
    </ul>
    
  );
}
