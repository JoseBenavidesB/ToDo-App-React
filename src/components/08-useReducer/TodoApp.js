
import React from 'react'; 
import { useReducer, useEffect } from 'react';
import { todoReducer } from './todoReducer';
//import { useForm } from '../../hooks/useForm';


import './styles.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';


const init = ()=> {

    return JSON.parse(localStorage.getItem( 'todos')) || [];
    /* return [{
        id: new Date().getTime(),
        desc: 'Learn React',
        done: false
    }]; */
}
export const TodoApp = () => {

    const [toDos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => { //save in the local storage
        
        localStorage.setItem( 'todos', JSON.stringify( toDos ) )
      }
    , [toDos]);
    
    const handleDelete = ( id ) => {
        const action = {
            type: 'delete',
            payload: id
        }
        //dispatch
        dispatch(action);
    };

    const handleToggle = ( todoId ) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    };
    const handleAddToDo = (newTodo) => {
        dispatch({
            type : 'add',
            payload : newTodo
        });
    };

    return (
        <>
            <h1> ToDo App ({ toDos.length })</h1> <hr/>

            <div key={1} className="row">
                <div className="col-7">

                    <TodoList 
                        toDos={ toDos } 
                        handleDelete={handleDelete} 
                        handleToggle={handleToggle} 
                    />

                </div >
                <div  className="col-5">
                    < AddTodo handleAddToDo={handleAddToDo}/>
                </div>
            </div>
            
        </>
    )
}
