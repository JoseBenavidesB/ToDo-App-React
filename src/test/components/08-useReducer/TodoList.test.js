import React from 'react';
import { shallow } from "enzyme/build"
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { newToDo } from '../../fixtures/demoToDo';

describe( 'Test TodoList', () => {
    
    const handleToggle = jest.fn(); 
    const handleDelete = jest.fn();

    const wrapper = shallow( <TodoList 
        toDos = {newToDo}
        handleToggle = { handleToggle }
        handleDelete = { handleDelete }
    />)

    test('should show TodoList correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should have two ToDo list item', () => {
        
        expect( wrapper.find( 'TodoListItem').length).toBe( newToDo.length );

        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function) );

        expect( wrapper.find('TodoListItem').at(0).prop('handleToggle') ).toEqual( expect.any(Function) );
    });


})