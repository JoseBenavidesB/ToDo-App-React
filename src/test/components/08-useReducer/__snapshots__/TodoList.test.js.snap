// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoList should show TodoList correctly 1`] = `
<ul
  className="list-group list-group-flush p-0"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    idx={0}
    key="1"
    todo={
      Object {
        "desc": "lear reac",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    idx={1}
    key="2"
    todo={
      Object {
        "desc": "lear mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
